### ISSUE TITLE
Upgrade notification-api to use Python 3.10

### ISSUE THREAD

**Issue by cris-oddball** [department-of-veterans-affairs/notification-api#1527]
# #1295 should be worked in parallel - there's details in this ticket.

## User Story - Business Need

We want to upgrade the notification-api to use Python 3.10, because 3.8 is nearing EOL in 2024, to maintain a secure and stable platform. In addition, we may not be able to upgrade the container to Alpine 3.18 (botocore/urllib3) without upgrading to Python 3.10 (latest botocore now supports urllib3 v2 but only for python 3.10+)

- [x] Ticket is understood, and QA has been contacted (if the ticket has a QA label).

### User Story(ies)

**As a** VANotify engineer
**I want** to upgrade the notification-api to use Python 3.10
**So that** the platform is secure , stable and supported.

## Engineering Checklist

- [x] Modify the docker container to use Python 3.10-alpine3.18
  - If this cannot be done: Document in a comment and try python3.10-alpine3.16
- [x] Locally, run the container and validate that the notification-api has no issues (many IDEs can report out incompatibilties)
- [x] Resolve any incompatibilities
  - [x] notification-utils has to work with 3.10
  - [x] dependencies that need to be upgraded
  - [x] classes/methods that may need to be modified to support 3.10
- [x] Modify any GitHub Actions that specifiy the python version related to API code
- [x] Deploy and ensure that the service works as expected
- [x] Update the readme with EOL for python 3.10

## Acceptance Criteria
GIVEN the container is using python 3.10
WHEN I run the notification-api service 
THEN the service performs as it did when using 3.8, including all regression tests passing

## QA Considerations
- [x] test test and test
- [x] templates with personalisation
- [x] testing for Twilio callbacks verified to work without error

## Out of Scope
Lambdas.

**Comment by mjones-oddball** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/VA-Notify-620d21369d810a00146ed9c8/issues/department-of-veterans-affairs/notification-api/1527?planning-poker) with Zenhub @cris-oddball @EvanParish @k-macmillan @kalbfled @ldraney @nikolai-efimov

**Comment by ldraney** [comment]
Team thoughts: Give this a shot with the current alpine image.  If there is an incompatibility, we can reapproach.

**Comment by ldraney** [comment]
More refinement thoughts: There is no perceived risk of the python version used in lambdas and this current ticket. 
We will update lambdas in a separate ticket.

**Comment by cris-oddball** [comment]
For DEV testing of personalisation

I created two templates in DEV for testing personalisation in the VANotify service.  There are a lot of templates in that service that use personalisation, so plenty of others to use for testing. Here are my notes for testing using the two I created. 

The identical templates currently work as expected in PERF., which does not yet have this code in it.

```
DEV VANotify Templates:
1	- "name": "Notify 1527 Python Upgrade Personalisation 20x",
	- "id": "9bcc03bf-2447-4d5e-a929-f95613624842",
	- properties are not in alphabetical order, to test, send a notification and set the personalisation properties in alphabetical order to its lower case version. When the email is received, it will be obvious whether the correct properties were populated or not.
2	- "name": "Notify 1527 Python Upgrade Personalisation All Same",
	- "id": "6481f991-b4b2-47be-a410-bd73c26fbfa7",
	- the same property is used throughout ((name)), send an email using this personalisation field multiple times.
```

@edDocMe360 @k-macmillan

**Comment by edDocMe360** [comment]
Updated the unit tests in utils to reflect the changes needed in utils to upgrade to 3.10. Deployed api to dev against these changes. Ran regression tests. 

https://github.com/department-of-veterans-affairs/notification-api-qa/actions/runs/7402769180

**Comment by cris-oddball** [comment]
This has two approvals and was merged. All testing note will be on the ticket, not the alpine one. Sending to perf for prolonged testing around templates.

Unfortunately Twistlock is still down, so cannot check what dependencies were cleared.

**Comment by cris-oddball** [comment]
Twistlock is up! Will be testing to see what exclusions are cleared with the upgrade to alpine 3.19 on that ticket.

**Comment by cris-oddball** [comment]
[Regression](https://github.com/department-of-veterans-affairs/notification-api-qa/actions/runs/7414471561) passes in Perf.

**Comment by cris-oddball** [comment]
Template testing

- [ ] TC01 send a template that has 20 personalisation fields, but send the POST request with the order different than what appears in the template. This payload and template are described above
- [ ] TC02 send a template that has 5 personalisation fields that are all the same property. This test is also described above.
- [ ] TC03 send a template that has a subject personalisation field and verify that the correct property appears in the subject `"id": "618185c6-3636-49cd-b7d2-6f6f5eb3bdde",`
- [ ] TC04 send a template that is in use in Perf with personalisation fields and verify it works `"id": "ab62e092-c9f4-48e2-b123-59abe23cc1f3",`

**Comment by cris-oddball** [comment]
Callback testing
Using Twilio testing service in Perf

- [ ] send a message through twilio with callback enabled and observe that the webhook is pinged
<img width="1117" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBL21kQWc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--1886e8f8a4916abc7657771a466cda9632ebd060/Screenshot_2024-01-04_at_3_15_07%E2%80%AFPM.jpg" alt="Screenshot_2024-01-04_at_3_15_07 PM.jpg" />

**Comment by cris-oddball** [comment]
Going to let this bake overnight to check BQ and logs for errors.

**Comment by cris-oddball** [comment]
Logs this morning do not show anything abnormal. BigQuery working as expected. Closing ticket.
